### PR TITLE
Build package for Pyodide out of tree

### PR DESCRIPTION
**Pull request by lagru**
## Description

Attempt to build our own wheel for the Pyodide distribution. If this proves "easy" we could test against it regularly and catch errors before releases. It would also make it easier to include JupyterLite in our HTML docs.

Related resources:

- Initial inspiration from https://github.com/numpy/numpy/pull/24603.
- [Building and testing Python packages out of tree](https://pyodide.org/en/stable/development/building-and-testing-packages.html)

cc @stefanv 

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

Summarize the introduced changes in the code block below in one or a few sentences. The
summary will be included in the next release notes automatically:

```release-note
...
```

### CODE REVIEW

File: .github/workflows/emscripten.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+          # that we can pass config-settings arguments to the backend.
+          # We'll then run into static assert issues in `pyconfig.h` due to:
+          # https://github.com/pyodide/pyodide/pull/2494#issuecomment-1118500743
+          # See also https://github.com/pyodide/pyodide/pull/2238 and the patches
+          # to _numpyconfig.h in https://github.com/pyodide/pyodide/tree/main/packages/numpy
+          #
+          # tl;dr this needs some work, and it seems like cross-compilation is
+          # hard because Pyodide doesn't ship its own Python interpreter that
+          # we can actually target during a direct cross build.
+          python -m build --wheel -Csetup-args=--cross-file=$PWD/tools/emscripten.meson.cross -Csetup-args=-Dallow-noblas=true
```
Comment (rgommers) [inline .github/workflows/emscripten.yml:67]: `-Csetup-args=-Dallow-noblas=true` should be removed, that is numpy-specific

File: .github/workflows/emscripten.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+          pyodide venv .venv-pyodide
+          source .venv-pyodide/bin/activate
+          pip install dist/*.whl
+          python -c "import sys; print(sys.platform)"
+          pip install -r test_requirements.txt
+      - name: Test
+        run: |
+          source .venv-pyodide/bin/activate
+          cd ..
+          pytest --pyargs numpy -m "not slow"
```
Comment (rgommers) [inline .github/workflows/emscripten.yml:85]: this should be `skimage` rather than `numpy`

File: tools/emscripten.meson.cross [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+sdl2-config = ['emconfigure', 'sdl2-config']
+
+[host_machine]
+system = 'emscripten'
+cpu_family = 'wasm32'
+cpu = 'wasm'
+endian = 'little'
+
+[properties]
+longdouble_format = 'IEEE_QUAD_LE'
```
Comment (rgommers) [inline tools/emscripten.meson.cross:20]: These two lines are numpy-specific, you can drop those.

**Comment by rgommers** [comment]
Nice to see this @lagru. I was just revisiting my own NumPy PR, and saw this one linked. Hopefully this is not too difficult to get over the finish line now that Pyodide has meson-python support.

**Comment by lagru** [comment]
@rgommers, thanks for taking a look! Now this is running into the same [`LONG_BIT` error](https://github.com/scikit-image/scikit-image/actions/runs/7180894298/job/19554131919?pr=7265#step:6:91) I see locally and mentioned in https://github.com/pyodide/pyodide/issues/4118. [You comment](https://github.com/numpy/numpy/pull/24603#issuecomment-1713509718) makes it seem that it may be a regression in the toolchain. I tried [pyodide-build 0.25.0a2](https://pypi.org/project/pyodide-build/0.25.0a2/) but that doesn't seem to make a difference. I'll keep looking but maybe the fix is obvious to you.

**Comment by rgommers** [comment]
Hmm, I was counting on that being resolved since the upstream bug was closed, and the fix confirmed to work for scikit-learn: https://github.com/pyodide/pyodide/pull/4136. That was with a Python from a conda env though, and that was the case for me as well (at least locally). Is it the same for you? If it's a different case, it may be worth commenting on https://github.com/pyodide/pyodide/issues/4118. 

I unfortunately didn't get time to dig deeper this week.

**Comment by stefanv** [comment]
I see the same problem, except that on Fedora `wordsize.h` is missing. But if I edit `pyconfig.h` to specifically include the 64-bit header, I replicate your problem:

```
$ cat /usr/include/python3.11/pyconfig.h
\\#include <bits/wordsize.h>

\\#if __WORDSIZE == 32
\\#include "pyconfig-32.h"
\\#elif __WORDSIZE == 64
#include "pyconfig-64.h"
\\#else
\\#error "Unknown word size"
\\#endif
```

I expect that pyodide is supposed to be looking inside the `.pyodide-xbuildenv` for header files, and not in `/usr/include`.

**Comment by stefanv** [comment]
I can add the include path needed:

```
python -m build --no-isolation --wheel -Csetup-args=--cross-file=/home/stefan/src/scikit-image/tools/emscripten.meson.cross -Csetup-args=--includedir=.pyodide-xbuildenv/xbuildenv/pyodide-root/cpython/installs/python-3.11.3/include/python3.11
```

But I do not know how to remove the standard Python include that is added.
My read of https://github.com/pyodide/pyodide/pull/4136/files is that those should be removed automatically, but :man_shrugging: this is all brand new to me :)

**Comment by lagru** [comment]
Thanks for investigating @stefanv. I just checked locally that this still seems to happen for the most recent pyodide-build versions 0.24.1 and 0.25.0a2 even though the fix https://github.com/pyodide/pyodide/pull/4136 should be included.

**Comment by lagru** [comment]
Ha, apparently there is no need to use a `meson.cross` file together with `python -m build ...` for pyodide 0.25.0a2: `pyodide build` "just works"! :tada: Now I only need to figure out how to run our test suite.

**Comment by lagru** [comment]
Okay, now I'm getting loads of `OSError: [Errno 8] Bad file descriptor` which I can reproduce locally as well.

**Comment by lagru** [comment]
Status update: it's probably reasonable to wait with this until after NumPy 2.0 has settled. See https://github.com/numpy/numpy/pull/24603#issuecomment-1937394674.

**Comment by agriyakhetarpal** [comment]
Hi there! I wanted to chime in into the conversation, since this was referenced in https://github.com/numpy/numpy/pull/24603 – its successor, here: https://github.com/numpy/numpy/pull/25894, should be merged soon (maybe during this week). Please let me know If there is any way @rgommers and I can drive this forward, we would be happy to help.

**Comment by lagru** [comment]
Thanks for keeping us up to date. I was quietly following along but I think it still might make sense to wait with this until Pyodide includes NumPy 2.0 officially? It's been some time but as I understand it the current approach with out-of-tree builds only targets one package, in this case scikit-image. So to build with NumPy 2.0 we'd have to wait until the Pyodide distribution includes this. Or we get NumPy 2.0 from somewhere else but that might be unnecessarily complicated when we can just wait? What do you think?

**Comment by rgommers** [comment]
> Okay, now I'm getting loads of `OSError: [Errno 8] Bad file descriptor` which I can reproduce locally as well.

We encountered this problem in PyWavelets. It was due to loading `.npy` files. The problem seems to be at least partially the same here; the last CI log shows a lot of failures for:

https://github.com/scikit-image/scikit-image/blob/47a627f1c9ea66486d05a65b06e618ebdb102fe3/skimage/morphology/footprints.py#L18

This was fixed a few days ago: https://github.com/pyodide/pyodide/issues/4541. The fix is not yet released, but it may be worth verifying that that takes care of all bad file descriptors with a local checkout of Pyodide (in case you don't have one at hand, maybe @agriyakhetarpal can give it a try?).

> I was quietly following along but I think it still might make sense to wait with this until Pyodide includes NumPy 2.0 officially? 

It is probably useful to not try to merge this PR before the needed `.npy` fix and other needed fixes are in a Pyodide release, rather than try to copy the `.patch` file from @agriyakhetarpal's NumPy PR. There isn't really anything numpy 2.0 specific here that I would expect to change though. Scikit-learn works with NumPy 1.26.x, so once this PR is green with that version I'd just merge it. When Pyodide upgrades to NumPy 2.0 (which may take several months), things should stay working.

**Comment by agriyakhetarpal** [comment]
> t may be worth verifying that that takes care of all bad file descriptors with a local checkout of Pyodide (in case you don't have one at hand, maybe @agriyakhetarpal can give it a try?).

> It is probably useful to not try to merge this PR before the needed .npy fix and other needed fixes are in a Pyodide release, rather than try to copy the .patch file from @agriyakhetarpal's NumPy PR.

I am happy to give this a try on a fork. However, the patch was for a different purpose – to allow NumPy to properly use `vendored-meson` though, isn't it? The easier solution to fix the failures here is to just convert all of the `.npy` files to the `.npz` file format, which were working even before https://github.com/pyodide/pyodide/issues/4541 was merged.

> There isn't really anything numpy 2.0 specific here that I would expect to change though. Scikit-learn works with NumPy 1.26.x, so once this PR is green with that version I'd just merge it. When Pyodide upgrades to NumPy 2.0 (which may take several months), things should stay working.

If it helps, we can keep this PR up to date before the imminent NumPy 2.0 release (or perhaps build these Pyodide/WASM wheels on both stable and development versions of NumPy – I'm not sure if that makes sense at this time). 

I would be glad to write a new PR that builds upon the changes here, or help out with code review here if needed. Please let me know what works!

**Comment by agriyakhetarpal** [comment]
In the meantime before this gets attention, I have tried updating the PR in this branch on my fork: [agriyakhetarpal/add-emscripten-builds-out-of-tree](https://github.com/agriyakhetarpal/scikit-image/tree/add-emscripten-builds-out-of-tree)

and owing to the fact that the patch for resolving the `.npy` file loading issue depends on another patch for Emscripten (it has to be backported because a new Emscripten release might take time), it's not very easy to apply it here – so, in https://github.com/scikit-image/scikit-image/commit/b5274dc165206b531a7a01545e9456e165a13e85 on this branch, I have converted all of the `.npy` files to `.npz` and updated their SHA-256 hashes just to test how the build goes – we can wait till a new release for Pyodide which should fix this change.

The package succeeds in compilation without any hiccups, but I received another failure during test collection due to `pooch` being unable to download the files via HTTP. I removed `pooch` from the workflow for now to proceed to use the data files (without checksum validation) from the local `skimage/data/` directory, where it now succeeds in collecting all test cases and fails after 65% of the test suite passes, due to an issue with unresolved symbols during linkage: https://github.com/agriyakhetarpal/scikit-image/actions/runs/8100301106/job/22138456770#step:9:263

which I am always unsure of as to how to debug them, but I did find this resource in the Pyodide issue tracker:

https://github.com/pyodide/pyodide/issues/3865#issuecomment-1559166778 but I do not think the discussion is related to what is happening here. I would appreciate any help I can on this!

**Comment by rgommers** [comment]
> I did find this resource in the Pyodide issue tracker: [pyodide/pyodide#3865 (comment)](https://github.com/pyodide/pyodide/issues/3865#issuecomment-1559166778)

That does look like exactly the same issue. And it doesn't look easy to work around. Let me add a comment on the upstream issue.

**Comment by lagru** [comment]
> I would be glad to write a new PR that builds upon the changes here, or help out with code review here if needed. Please let me know what works!

@agriyakhetarpal, please by all means feel very welcome to go ahead and take over. :pray: Re the NumPy 1 vs 2 issue, I think it's fine to just use the most recent NumPy version provided by the Pyodide distribution. Right now we are mainly interested to run our test suite with Pyodide and add JupyterLite support in our dev docs.

**Comment by agriyakhetarpal** [comment]
> @agriyakhetarpal, please by all means feel very welcome to go ahead and take over. 🙏 Re the NumPy 1 vs 2 issue, I think it's fine to just use the most recent NumPy version provided by the Pyodide distribution. Right now we are mainly interested to run our test suite with Pyodide and add JupyterLite support in our dev docs.

Thank you for the response, @lagru. I had put this aside for a while after receiving the above unresolved symbol `s_cmp` error, but I am happy to restart debugging and work towards a solution. A potential solution that @rgommers and I had thought of—which had also come up in our discussion while doing the out-of-tree builds for Matplotlib recently—was to test `scikit-image` inside a Node.js server instead of a Pyodide-managed virtual environment; it may not fix the issue at hand, but it is still worth trying out. I will open up a draft PR as soon as possible for visibility-related purposes, to supersede this one.

**Comment by lagru** [comment]
Superseded by https://github.com/scikit-image/scikit-image/pull/7350.

**Pull request closed without merging**